### PR TITLE
dotCMS/core#15898 SPAEM: Create react carousel component

### PR DESCRIPTION
**Pull request by alfredo-dotcms**

### CODE REVIEW

File: src/Components/BannerCarousel/BannerCarousel.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import Parser from 'html-react-parser';
+import { Carousel, CarouselItem, CarouselControl } from 'reactstrap';
+
+class BannerCarousel extends Component {
+    constructor(props) {
+        super(props);
+        this.state = { activeIndex: 0 };
+        this.next = this.next.bind(this);
```
Comment (fmontes) [inline src/Components/BannerCarousel/BannerCarousel.js]: You don't need the bind.

File: src/Components/BannerCarousel/BannerCarousel.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                <CarouselItem
+                    onExiting={this.onExiting}
+                    onExited={this.onExited}
+                    key={item.src}
+                >
+                    <img src={item.src} alt={item.altText} />
+                    {showTitle || showCaption ? (
+                        <div className="carousel-caption d-none d-md-block">
+                            {showTitle ? <h1>{item.title}</h1> : ''}
+                            {showCaption ? Parser(item.caption) : ''}
```
Comment (fmontes) [inline src/Components/BannerCarousel/BannerCarousel.js]: https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml

File: src/Components/BannerCarousel/BannerCarouselWidgets.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        this.getBannersItems();
+    }
+
+    render() {
+        const { loading, banners } = this.state;
+        if (loading) {
+            return <Spinner color="primary" />;
+        } else {
+            return (
+                <>
```
Comment (fmontes) [inline src/Components/BannerCarousel/BannerCarouselWidgets.js]: Don't need.

File: src/Components/BannerCarousel/BannerCarouselWidgets.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                    {banners && banners.length ? (
+                        <>
```
Comment (fmontes) [inline src/Components/BannerCarousel/BannerCarouselWidgets.js]: Don't need.

File: src/Components/Contentlet.js [hunk 5]
```diff
@@ -33,15 +34,17 @@ const SimpleWidget = ({data}) => {
 }
 
 const getComponent = (data) => {
-    switch (data.contentType) {
-        case 'webPageContent':
+    switch (data.contentType.toUpperCase()) {
```
Comment (fmontes) [inline src/Components/Contentlet.js]: Why?

File: src/Components/News/NewsDetail.js [hunk 6]
```diff
@@ -8,13 +9,18 @@ const NewsDetail = ({ news }) => {
             <h3>{title}</h3>
             <h5>{sysPublishDate}</h5>
             <img alt="" src={image} />
-            <p>{story}</p>
+            <div>{Parser(story)}</div>
```
Comment (fmontes) [inline src/Components/News/NewsDetail.js]: https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml

File: src/Components/News/NewsList.js [hunk 7]
```diff
@@ -19,7 +19,15 @@ const NewsList = ({ news, fieldsToDisplay }) => {
 
 NewsList.propTypes = {
     fieldsToDisplay: PropTypes.string.isRequired,
-    news: PropTypes.array.isRequired
+    news: PropTypes.arrayOf(
+        PropTypes.shape({
```
Comment (fmontes) [inline src/Components/News/NewsList.js]: NICEEEE!!! I was looking forward do type this.

File: src/Components/News/NewsListItem.js [hunk 8]
```diff
@@ -36,7 +36,14 @@ const NewsListItem = ({ news, fieldsToDisplay }) => {
 
 NewsListItem.propTypes = {
     fieldsToDisplay: PropTypes.string,
-    news: PropTypes.object.isRequired
+    news: PropTypes.shape({
```
Comment (fmontes) [inline src/Components/News/NewsListItem.js]: ❤️

File: src/Components/News/NewsWidget.js [hunk 9]
```diff
@@ -74,7 +74,7 @@ class NewsWidget extends React.Component {
                 <>
                     {news && news.length ? (
                         <>
-                            <NewsList news={news} fieldsToDisplay={fieldsToDisplay} />
+                            <NewsList news={news} fieldsToDisplay={fieldsToDisplay || ''} />
```
Comment (fmontes) [inline src/Components/News/NewsWidget.js]: Why? could this be null or something?

File: src/__tests__/BannerCarousel/BannerCarouselWidgets.test.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            esresponse: [
+                {
+                    hits: {
+                        total: 3
+                    }
+                }
+            ]
+        };
+
+        DotCMSApi.esSearch = jest.fn().mockImplementation(() => {
```
Comment (fmontes) [inline src/__tests__/BannerCarousel/BannerCarouselWidgets.test.js]: DRY: you can extract this and pass the response and return the mock.

File: src/Components/BannerCarousel/BannerCarousel.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    goToIndex(newIndex) {
+        if (this.animating) return;
+        this.setState({ activeIndex: newIndex });
+    }
+
+    render() {
+        const displayedFields = this.props.fieldsToDisplay.split(',');
+        const { activeIndex } = this.state;
+        const slides = this.props.items.map(item => {
+            const showTitle = displayedFields.includes('title');
```
Comment (fmontes) [inline src/Components/BannerCarousel/BannerCarousel.js]: Why do this in each iteration?

File: src/Components/BannerCarousel/BannerCarousel.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            const showCaption = displayedFields.includes('caption1');
+            return (
+                <CarouselItem
+                    onExiting={this.onExiting}
+                    onExited={this.onExited}
+                    key={item.src}
+                >
+                    <img src={item.src} alt={item.altText} />
+                    {showTitle || showCaption ? (
+                        <div className="carousel-caption d-none d-md-block">
```
Comment (fmontes) [inline src/Components/BannerCarousel/BannerCarousel.js]: There is a `CarouselCaption` reactstrap component.

File: src/Components/BannerCarousel/BannerCarouselWidgets.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            sortResultsBy: sortResultsByMap[this.props.data.sortResultsBy]
+        };
+
+        DotCMSApi.esSearch('banner', defaultEsParams)
+            .then(response => response.json())
+            .then(bannerData => {
+                this.setState(state => ({
+                    ...state,
+                    loading: false,
+                    banners: this.formatBannersData(bannerData.contentlets)
```
Comment (fmontes) [inline src/Components/BannerCarousel/BannerCarouselWidgets.js]: If you like... you could do this in a `then`... just saying.

File: src/Components/BannerCarousel/BannerCarousel.js (hunk 2 above)
Comment (alfredo-dotcms) [reply src/Components/BannerCarousel/BannerCarousel.js]: I used it, so I didn't have to do any data sanitation:
https://stackoverflow.com/questions/29044518/safe-alternative-to-dangerouslysetinnerhtml

File: src/Components/BannerCarousel/BannerCarousel.js (hunk 12 above)
Comment (alfredo-dotcms) [reply src/Components/BannerCarousel/BannerCarousel.js]: this only accepts `strings` and the value from the `BE` contains html tags/text

File: src/Components/BannerCarousel/BannerCarousel.js (hunk 2 above)
Comment (fmontes) [reply src/Components/BannerCarousel/BannerCarousel.js]: But also it says:

> I figure, if you trust your own CMS/Server (and not receiving from a 3rd party), which has sanitized the data (this step is also done), then you can insert using dangerouslySetInnerHTML

In this case, we just want to deliver this as "native" react we can, the less amount of extra npm module as we can.

**Pull request merged**